subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java

```diff
@@ -196,80 +254,64 @@ public void afterExecute(Task task, TaskState state) {
         });
     }
 
+    @Override
     public boolean hasTask(Task task) {
-        ensurePopulated();
-        return taskExecutionPlan.getTasks().contains(task);
+        return getWorkGraph().hasTask(task);
```

Comment:
Doesn't the `workGraph` change while we are executing the build? I think we are removing the nodes from the work graph when they finish.

Reply:
Yes, but they remain in the `taskNodes` map.